src/ape/utils/misc.py

```diff
@@ -230,6 +231,26 @@ def inner(*args, **kwargs):
     return inner
 
 
+class cached_iterator(property):
```

Comment:
@fubuloubu I used your `itertools.tee()` recommendation in combo with the UX of `@cached_property` so now you can do

```python
@cached_iterator
def trace() -> Iterator[TraceFrame]
   ..
```

and it only makes the RPC once.
I have tests on the hardhat plugin for this that I will push up tomorrow.

Reply:
You van probably unit test this impl directly and then we can merge this PR nowish